Return specific AZ's for taskcat_getsingleaz

bmoller opened the pull request:
Looking at a few different repos, it appears that the expected functionality of the `taskcat_getsingleaz` substitution differs from its actual implementation. Currently this will always return a single AZ as expected, but always the first AZ as returned from the API lookup.

This change allows template parameters to select specific AZ's by index, which seems to match the intended usage.

## Overview

It appears that this was intended to allow a template's parameters to select availability zones without knowing what they are. CloudFormation provides this as a pseudo-parameter mapping, but it may make more sense to allow users to specify availability zones at launch time. In this case, taskcat needs a similar kind of mapping lookup.

For multiple examples, see the various compliance-* repos. Here's the PCI compliance Quick Start's parameters: https://github.com/aws-quickstart/quickstart-compliance-pci/blob/master/ci/defaults.json

## Testing/Steps taken to ensure quality

Tested locally using templates with substitution's similar to `taskcat_getsingleaz_1`, `taskcat_getsingleaz_2`, etc.

## Testing Instructions

Test 1:
  * Create taskcat `defaults.json` parameters file that uses the `$[taskcat_getsingleaz_1]` and `$[taskcat_getsingleaz_2]` substitutions.
  * Run taskcat and check that the template parameters are properly filled in with the first and second AZ's from the region

Test 2:
  * Try using an invalid index for the AZ, such as `$[taskcat_getsingleaz_100]`
  * Run taskcat and check that it exits with an error stating that there are not that many AZ's in the region

bmoller commented:
Addresses issue #218

jaymccon commented:
@agamzn this will likely cause some conflicts in your upcoming pr https://github.com/agamzn/taskcat/tree/issues/149, be sure to rebase and check tests reflect this change

jaymccon commented on taskcat/stacker.py:
```diff
@@ -986,12 +1007,19 @@ def generate_input_param_values(self, s_parms, region):
+                        print(PrintMsg.INFO + "$[taskcat_getsingleaz_(!)]")
+                        print(PrintMsg.INFO + "Which az not specified!")
+                        print(PrintMsg.INFO + " - (Defaulting to az 1)")
+                        param_value = self.get_single_az(region, 1)
```
might be nice to return a random az in this case ?

bmoller replied:
No longer applicable, per our discussion below and the latest push.

jaymccon commented:
Your second test `$[taskcat_getsingleaz_100]` doesn't match the regex, as the az number is expected to be single digit `\$\[\w+_ge[nt]singleaz_\d]`. Providing `$[taskcat_getsingleaz_9]` causes taskcat to `exit 1`, which is inconsistent with `$[taskcat_getsingleaz_100]` which just carries on without replacing the target string.

Not sure what the actual desired behavior is in this case, but it seems that both of those may be undesired. Thinking of the case where we want to support regions with both 2 and 3 az's. For a region with 2 az's `$[taskcat_getsingleaz_3]` will cause taskcat to fail. Might be a better behavior to just return some az. That said, I can't think of an actual quick start that needs this flexibility, so at minimum the behavior for referencing a non-existent az number should be consistent from 3-∞.

jaymccon commented on taskcat/stacker.py:
```diff
@@ -986,12 +1007,19 @@ def generate_input_param_values(self, s_parms, region):
+                        print(PrintMsg.DEBUG + "Selecting availability zone %s" % az_id)
+                        param_value = self.get_single_az(region, az_id)
+                        _parameters['ParameterValue'] = param_value
+                    else:
```
this code path will only execute when `az_id == 0`, is this intended ?

bmoller replied:
Yeah, I caught on to the same thing and removed the construct in the updated push. The same holds true for the `$[taskcat_genaz_\d]` regex; it only matches if there's a 0 there. What's the impact in that case?

bmoller commented:
I'm too new to have an opinion as to what the "correct" behavior should be at higher numbers; whatever the consensus is I'll implement.

I definitely think taskcat should pick up something like `$[taskcat_getsingleaz_9999999]` and do something, but I'll defer to wiser and more senior opinions as to what that should be. Being relatively new, I would expect it to return and quit with an error. I don't like when tools try and do the right thing and would prefer that my error be called out.

jaymccon commented on taskcat/stacker.py:
```diff
@@ -490,6 +490,27 @@ def get_available_azs(self, region, count):
+        :return: The requested availability zone of the specified region.
+        """
+
+        regional_azs = self.get_available_azs(region, az_id)
```
what's the benefit of the extra abstraction here over calling `ec2_client.describe_availibility_zones` directly?

bmoller replied:
My goal was to add as little as possible. I'm still new to this codebase and wanted to make as small and easily understood an impact as possible.

jaymccon commented:
> I would expect it to return and quit with an error. I don't like when tools try and do the right thing and would prefer that my error be called out.

👍 to this approach

@avattathil @sshvans @vsnyc anyone want to weigh in on this ?

santiagocardenas commented:
For what it’s worth, this generator used to work in Alfred.

bmoller commented:
I updated the regex to catch any `int` with 3+ places. Also realized the `if ... else` block was useless and removed it.

Also, as a sidenote, I couldn't help but notice that the `$[taskcat_genaz_\d]` construct [has the same irregularity](https://github.com/aws-quickstart/taskcat/blob/develop/taskcat/stacker.py#L967-L986). The regex requires that a digit be there, so the `else` will never execute. Happy to clean that up as well if it makes sense.

jaymccon commented:
> Also, as a sidenote, I couldn't help but notice that the `$[taskcat_genaz_\d]` construct [has the same irregularity](https://github.com/aws-quickstart/taskcat/blob/develop/taskcat/stacker.py#L967-L986). The regex requires that a digit be there, so the `else` will never execute. Happy to clean that up as well if it makes sense.

Would be great if you could address in a separate pr. Don't want to hold this one up.
